dev-python/cadquery-ocp/cadquery-ocp-7.5.2_rc20210706.ebuild

```diff
@@ -0,0 +1,160 @@
+	local _OCCT_LIB_DIR="${EPREFIX}/usr/lib64/opencascade-${OCCT_PV}"
+
+	# Absolute dirname of VTK's include directory,
+	local _VTK_INCLUDE_DIR="${EPREFIX}/usr/include/vtk-${_VTK_VERSION}"
+
+	# Ensure the above paths exist as a crude sanity test.
+	test -d "${_CLANG_INCLUDE_DIR}" || die "${_CLANG_INCLUDE_DIR} not found."
+	test -f "${_CLANG_LIB_FILE}"    || die "${_CLANG_LIB_FILE} not found."
+	test -d "${_OCCT_INCLUDE_DIR}"  || die "${_OCCT_INCLUDE_DIR} not found."
+	test -d "${_OCCT_LIB_DIR}"      || die "${_OCCT_LIB_DIR} not found."
+	test -d "${_VTK_INCLUDE_DIR}"   || die "${_VTK_INCLUDE_DIR} not found."
+
+	# "dev-python/clang-python" atom targeting this Clang version.
+	local _CLANG_PYTHON_ATOM="dev-python/clang-python-${_CLANG_VERSION}"
+
+	# Ensure "dev-python/clang-python" targets this Clang version.
+	has_version -r "=${_CLANG_PYTHON_ATOM}" ||
+		die "${_CLANG_PYTHON_ATOM} not installed."
+
+	# Remove all vendored subdirectories.
+	rm -rf conda opencascade pywrap || die
+
+	# Inject a symlink to OCCT's include directory.
+	ln -s "${_OCCT_INCLUDE_DIR}" opencascade || die
+
+	python_copy_sources
+	cadquery-ocp_src_prepare() {
+		cd "${BUILD_DIR}" || die
+
+		# Generate OCCT bindings in the "${BUILD_DIR}/OCP/" subdirectory.
+		${EPYTHON} -m bindgen \
+			--verbose \
+			--njobs $(get_nproc) \
+			--libclang "${_CLANG_LIB_FILE}" \
+			--include "${_CLANG_INCLUDE_DIR}" \
+			--include "${_VTK_INCLUDE_DIR}" \
+			all ocp.toml || die
+
+		# Remove the source "FindOpenCascade.cmake" after generating bindings,
+		# which copied that file to the target "OCP/FindOpenCascade.cmake".
+		rm FindOpenCascade.cmake || die
+
+		#FIXME: Submit an upstream issue. This is frankly awful.
+		# Replace all hardcoded paths in "OCP/FindOpenCascade.cmake" with
+		# standard OCCT paths derived above. That file is both fundamentally
+		# broken and useless, as the ${CASROOT} environment variable and
+		# "/usr/lib64/cmake/opencascade-${PV}/OpenCASCADEConfig.cmake" file
+		# already reliably identify all requisite OpenCASCADE paths. Failure to
+		# patch this file results in src_configure() failures resembling:
+		#     -- Could NOT find OPENCASCADE (missing: OPENCASCADE_LIBRARIES)
+		sed -i \
+			-e 's~$ENV{CONDA_PREFIX}/include/opencascade\b~'${_OCCT_INCLUDE_DIR}'~' \
+			-e 's~$ENV{CONDA_PREFIX}/lib\b~'${_OCCT_LIB_DIR}'~' \
+			-e 's~$ENV{CONDA_PREFIX}/Library/\(lib\|include/opencascade\)~~' \
+			OCP/FindOpenCascade.cmake || die
+	}
+	python_foreach_impl cadquery-ocp_src_prepare
+}
+
+src_configure() {
+	cadquery-ocp_src_configure() {
+		cd "${BUILD_DIR}" || die
+		cmake \
+			-Wno-dev \
+			-S OCP/ \
+			-B OCP.binary_tree/ \
+			-G Ninja \
+			-D CMAKE_BUILD_TYPE=Gentoo \
+			-D PYTHON_EXECUTABLE="${PYTHON}" \
+			|| die
+	}
+	python_foreach_impl cadquery-ocp_src_configure
+}
+
+src_compile() {
+	cadquery-ocp_src_compile() {
+		cd "${BUILD_DIR}" || die
+		cmake --build OCP.binary_tree/ -- -j $(get_nproc) || die
```

Comment:
Shouldn't we honour the users choice by using `makeopts_jobs` instead of `get_nproc`?

Reply:
Somewhat related to this, I have a feeling that major parts of the emerge are not parallelized at all:
```
* dev-python/cadquery-ocp-7.5.2_rc20210706

current merge time: 5 hours, 7 minutes and 58 seconds.
ETA: any time now.
```
And it's still busy.....

The System Monitor shows me that it is clearly using only one core at the moment (I haven't been monitoring it all the time, so maybe it was multithreaded before)

If we can't speed this up by parallelizing it (more), then maybe we should use `python-single-r1` instead of `python-r1`. Because if this really takes this long than no one is going to want to install this for multiple PYTHON_TARGETS since that doubles or triples the emerge time. (This also has the additional benefit of simplifying the cmake problem, since we don't have to deal with multiple BUILDDIRs)

Reply:
> Shouldn't we honour the users choice by using `makeopts_jobs` instead of `get_nproc`?

...yes. Yes, we should. I actually didn't know that `makeopts_jobs()` even existed, because I never actually read the eclass documentation. Let's pretend I didn't just admit that.

Reply:
> I have a feeling that major parts of the emerge are not parallelized at all...

That's... not good. I thought it was just me. I'm relieved to learn that my geriatric CPU isn't entirely too blame, but disappointed to learn that OCP may have no idea what it's doing. In theory:

* The `src_prepare()` phase *should* be parallelized, because we explicitly pass `--njobs $(makeopts_jobs)`. This assumes Cadquery pywrap works as expected, which may be too much to ask.
* The `src_compile()` phase *should* be parallelized too, both because we explicitly pass `-j $(makeopts_jobs)` to `ninja` *and* because `ninja` defaults to parallelized compilation dependent on core count + 1.
* The `src_configure()` phase doesn't appear to be parallelized but probably should be.

Let's see if I can track down exactly what isn't happening and why. `<cracks_arthritic_knuckles>`

Reply:
I can confirm that `src_prepare()` is parallelized. So, that's nice.

I suspect we're just going to discover that generating Python bindings to a 1.5GB C++ API unavoidably consumes a *crazy* amount of wallclock time. I'm kinda okay with that. This is the nature of the beast. Of course, half-a-day build times makes this more a sloth than a beast.

Still, I'd kinda be nice if we could preserve multiple `PYTHON_TARGETS`. Gentoo is choice, even if that means a bad choice. More relevantly, CadQuery will probably end up becoming a mandatory runtime dependency of various other downstream Python projects in a few years. Dropping multiple targets for CadQuery means dropping multiple targets for those future packages, too.

Users on painfully outdated systems that wheeze uncontrollably (like my own) can always disable multiple targets with USE flags like:

```
dev-python/cadquery-ocp -python_targets_python3_8
```

Maybe we could emit a warning message in `pkg_setup()` when multiple targets are enabled?

Reply:
> Maybe we could emit a warning message in `pkg_setup()` when multiple targets are enabled?

That would work too, it might even be better because then the user is warned about what they are getting into when compiling this.

I haven't checked the memory/disk usage, but it is probably large too. So we might want to set `CHECKREQS_DISK_BUILD` from `check-reqs.eclass`. And we should probably set it conditionally to how many `PYTHON_TARGETS` are enabled, something like:
```
pre_build_checks() {
     local M=0
     for x in ${PYTHON_COMPAT}; do
         M=$(( <some value for each target in Mb> + ${M} ))
    done
    CHECKREQS_DISK_BUILD=${M}M check-reqs_pkg_${EBUILD_PHASE}
}

pkg_pretend() {
	pre_build_checks
}

pkg_setup() {
	pre_build_checks
}
```
(untested, and stolen from rust.ebuild)

Reply:
I approve of ~~stealing~~ being inspired by everything not nailed down from Rust.

My crude and probably wrong disk usage heuristic is showing ~400MB disk consumption and negligible memory consumption (...maybe because parsing is much less memory intensive than linking?) per Python target. Your hunch was spot on; the disk consumption is severe enough that we really do want to enforce pre-build constraints. Let's do this!

Reply:
I just discovered this little helper function that might come in handy here:
```
ebegin

Like einfo, we output a helpful message and then hint that the following operation may take some time to complete. Once the task is finished, you need to call eend. 
```